### PR TITLE
Fix for 20303. 

### LINKED ISSUE #20303: Throw early when user mistakenly uses OwnsOne() but points to a collection
If the user mistakenly uses `OwnsOne()` on a collection or `OwnsMany()` on a reference we should point this out with an error message.

E.g. see https://github.com/dotnet/efcore/blob/112b6064fae7f4fd61af65c8dee16e711734a966/test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs#L1394

If you change that line to an `OwnsOne()` call instead, the `NonGenericString` version of the test eventually fails but with an unhelpful error message deep in the code. Some of the other (`Generic`/`NonGeneric` etc) OwnedTypes Navigation tests actually pass.

We should throw in that situation providing a helpful error message as early as possible.

### PR DESCRIPTION
**Pull request by lajones**
Throw if cannot set IsUnique because the NavProp has the opposite multiplicity.

Fixes #20303.

### CODE REVIEW

File: test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs [hunk 1]
```diff
@@ -1437,6 +1437,24 @@ public virtual void Navigations_on_OwnsMany_Owned_types_can_set_access_mode_usin
...
+
+                Assert.Equal(
+                    CoreStrings.UsedOwnsManyOnReference(
+                        "OwnedDependent",
+                        typeof(OneToOneNavPrincipalOwner).FullName,
+                        typeof(OwnedNavDependent).FullName),
+                    Assert.Throws<InvalidOperationException>(
+                        () => modelBuilder
+                            .Entity<OneToOneNavPrincipalOwner>()
+                            .OwnsMany<OwnedNavDependent>("OwnedDependent")).Message
```
Comment (AndriySvyryd) [inline test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs]: Does this throw without your fix?

Comment (lajones) [reply test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs]: No.

Comment (AndriySvyryd) [reply test/EFCore.Tests/ModelBuilding/OwnedTypesTestBase.cs]: Then the fix should be in https://github.com/dotnet/efcore/blob/c42afa1caac8cf26ddd91d6f0cdf74301ca4b4c0/src/EFCore/Metadata/Internal/InternalForeignKeyBuilder.cs#L1395
We shouldn't remove a explicitly-configured navigation silently

File: src/EFCore/Metadata/Builders/EntityTypeBuilder.cs [hunk 2]
```diff
@@ -563,6 +566,42 @@ private OwnedNavigationBuilder OwnsManyBuilder(in TypeIdentity ownedType, string
...
+            {
+                return;
+            }
+
+            var enumerableOfTargetEntityTypeClrType =
+                typeof(IEnumerable<>).MakeGenericType(targetEntityTypeClrType);
+            var navigationPropertyClrType =
+                foreignKeyBuilder.Metadata.PrincipalToDependent?.ClrType;
+            if (navigationPropertyClrType != null
+                && !enumerableOfTargetEntityTypeClrType.IsAssignableFrom(navigationPropertyClrType))
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Builders/EntityTypeBuilder.cs:595]: Use `navigationPropertyClrType?.TryGetSequenceType()` instead

**Comment by lajones** [comment]
@AndriySvyryd Completely updated this PR to take the new approach.

**Pull request merged**